src/main/java/org/opensearch/security/transport/SecurityRequestHandler.java

```diff
@@ -113,16 +113,28 @@ protected void messageReceivedDecorate(
 
         String initialActionClassValue = getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_INITIAL_ACTION_CLASS_HEADER);
 
-        String userHdr = getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER_HEADER);
-
-        // restore a persistent user-subject from header, if null
-        if (getThreadContext().getPersistent(ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER) == null && userHdr != null) {
-            User user = this.userFactory.fromSerializedBase64(userHdr);
+        String authUsrHdr = getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER_HEADER);
+        String shouldUseUserHeader = getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_SAME_AS_SUBJECT_HEADER);
+        String userHdr = getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER);
+
+        // restore a persistent user-subject from subject header
+        if (getThreadContext().getPersistent(ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER) == null) {
+            // when auth subject user is same request user.
+            if (Boolean.parseBoolean(shouldUseUserHeader) && userHdr != null) {
+                User user = this.userFactory.fromSerializedBase64(userHdr);
+
+                getThreadContext().putPersistent(
+                    ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER,
+                    new UserSubjectImpl(getThreadPool(), user)
+                );
+            } else if (authUsrHdr != null) {
+                User user = this.userFactory.fromSerializedBase64(authUsrHdr);
```

Comment:
The original user handling is done in these lines:

https://github.com/opensearch-project/security/blob/46e9e2bb23cc4824d06c0e034121ca0bad27791f/src/main/java/org/opensearch/security/transport/SecurityRequestHandler.java#L197-L214

Can't we combine these into one? With the current code, we parse the user header two times

Reply:
no because line 141 calls newStoredContext and the else block is only hit for netty requests. Adding this there makes code look complicated. Also, reading from header should be okay since its a map and should be a O(1) operation.

Reply:
done [5d41f95](https://github.com/opensearch-project/security/pull/5408/commits/5d41f95a2ba8969afe90ccb4a0bce52d16672108)